File: library/src/main/java/name/zeno/android/widget/ZTextWatcher.java
package name.zeno.android.widget;

import android.support.annotation.CallSuper;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;
import android.widget.TextView;

import java.util.regex.Pattern;

import name.zeno.android.listener.Action2;

/**
 * Create Date: 16/5/30
 *
 * @author 陈治谋 (513500085@qq.com)
 */
public abstract class ZTextWatcher implements TextWatcher
{

  public static TextWatcher watch(final TextView view, final Action2<TextView, String> watcher)
  {
    TextWatcher textWatcher = new TextWatcher()
    {
      private TextView textView = view;

      @Override public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }

      @Override public void onTextChanged(CharSequence charSequence, int i, int i1, int i2)
      {
        if (watcher != null) {
          watcher.call(textView, charSequence.toString());
        }
      }

      @Override public void afterTextChanged(Editable editable) { }
    };

    view.addTextChangedListener(textWatcher);
    return textWatcher;
  }

  /**
   * 通过 {@link TextWatcher} 纠正输入
   */
  public static TextWatcher pattern(final EditText view, String regex)
  {
    TextWatcher textWatcher = new TextWatcher()
    {
      Pattern pattern = Pattern.compile(regex);
      private EditText textView = view;
      CharSequence old = "";

      @Override public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2)
      {
        if (old.length() == 0 || pattern.matcher(charSequence).matches()) {
          old = charSequence.toString();
        }
      }

      @Override public void onTextChanged(CharSequence s, int i, int i1, int i2)
      {
        if (s.length() > 0 && !old.equals(s.toString()) && !pattern.matcher(s).matches()) {
          int p = textView.getSelectionStart();
          p += s.length() > old.length() ? -1 : +1;
          textView.setText(old);
          textView.setSelection(Math.min(p, textView.length()));
        }
      }

      @Override public void afterTextChanged(Editable editable) { }
    };

    view.addTextChangedListener(textWatcher);
    return textWatcher;
  }

  @Override
  public void afterTextChanged(Editable s) { }

  @Override
  public void beforeTextChanged(CharSequence s, int start, int count, int after) { }

  @CallSuper
  @Override
  public void onTextChanged(CharSequence s, int start, int before, int count)
  {
    onTextChanged(s.toString());
  }

  abstract public void onTextChanged(String txt);

}

File: library/src/main/java/name/zeno/android/widget/ZTextWatcher.kt
package name.zeno.android.widget

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import android.widget.TextView
import androidx.annotation.CallSuper
import java.util.regex.Pattern

/**
 * Create Date: 16/5/30
 *
 * @author 陈治谋 (513500085@qq.com)
 */
abstract class ZTextWatcher : TextWatcher {

  override fun afterTextChanged(s: Editable) {}

  override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

  @CallSuper
  override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
    onTextChanged(s.toString())
  }

  abstract fun onTextChanged(txt: String)

  companion object {

    fun watch(view: TextView, watcher: (view: TextView, text: String) -> Unit): TextWatcher {
      val textWatcher = object : TextWatcher {

        override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

        override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
          watcher(view, charSequence.toString())
        }

        override fun afterTextChanged(editable: Editable) {}
      }

      view.addTextChangedListener(textWatcher)
      return textWatcher
    }

    /**
     * 通过 [TextWatcher] 纠正输入
     */
    fun pattern(view: EditText, regex: String): TextWatcher {
      val textWatcher = object : TextWatcher {
        internal var pattern = Pattern.compile(regex)
        internal var old: CharSequence = ""

        override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
          if (old.isEmpty() || pattern.matcher(charSequence).matches()) {
            old = charSequence.toString()
          }
        }

        override fun onTextChanged(s: CharSequence, i: Int, i1: Int, i2: Int) {
          if (s.length > 0 && old != s.toString() && !pattern.matcher(s).matches()) {
            var p = view.selectionStart
            p += if (s.length > old.length) -1 else +1
            view.setText(old)
            view.setSelection(Math.min(p, view.length()))
          }
        }

        override fun afterTextChanged(editable: Editable) {}
      }

      view.addTextChangedListener(textWatcher)
      return textWatcher
    }
  }

}
